src/Libraries/Microsoft.Extensions.TimeProvider.Testing/FakeTimeProvider.cs

```diff
@@ -12,83 +12,109 @@
 
     /// <summary>
     /// Initializes a new instance of the <see cref="FakeTimeProvider"/> class.
     /// </summary>
-    /// <param name="epoch">The starting point for the clock used by this <see cref="FakeTimeProvider"/>.</param>
-    [Experimental]
-    public FakeTimeProvider(DateTimeOffset epoch)
-        : this()
+    /// <param name="startTime">The initial time and date reported by the provider.</param>
+    /// <param name="autoAdvanceAmount">The amount of time by which to automatically advance time whenever it is read.</param>
+    public FakeTimeProvider(DateTimeOffset startTime, TimeSpan autoAdvanceAmount)
     {
-        Epoch = epoch;
-        _now = epoch;
+        _ = Throw.IfLessThan(autoAdvanceAmount.Ticks, 0);
+
+        _now = startTime;
+        _autoAdvanceAmount = autoAdvanceAmount;
     }
 
     /// <inheritdoc />
     public override DateTimeOffset GetUtcNow()
     {
-        return _now;
+        List<FakeTimeProviderTimer.Waiter>? waitersToWake;
+        DateTimeOffset result;
+
+        lock (Waiters)
+        {
+            result = _now;
+            _now += _autoAdvanceAmount;
+            waitersToWake = GetWaitersToWake();
+        }
+
+        WakeWaiters(waitersToWake);
+        return result;
     }
 
     /// <summary>
-    /// Sets the date and time in the UTC timezone.
+    /// Sets the date and time in the UTC time zone.
     /// </summary>
-    /// <param name="value">The date and time in the UTC timezone.</param>
+    /// <param name="value">The date and time in the UTC time zone.</param>
     public void SetUtcNow(DateTimeOffset value)
     {
-        if (value < _now)
-        {
-            Throw.ArgumentOutOfRangeException(nameof(value), $"Cannot go back in time. Current time is {GetUtcNow()}.");
-        }
+        List<FakeTimeProviderTimer.Waiter>? waitersToWake = null;
 
-        while (value >= _now && TryGetWaiterToWake(value) is FakeTimeProviderTimer.Waiter waiter)
+        lock (Waiters)
         {
-            _now = waiter.WakeupTime;
-            waiter.TriggerAndSchedule(false);
+            if (value < _now)
+            {
+                Throw.ArgumentOutOfRangeException(nameof(value), $"Cannot go back in time. Current time is {_now}.");
+            }
+
+            _now = value;
+            waitersToWake = GetWaitersToWake();
         }
 
-        _now = value;
+        WakeWaiters(waitersToWake);
     }
 
     /// <summary>
-    /// Advances the clock's time by a specific amount.
+    /// Advances time by a specific amount.
     /// </summary>
     /// <param name="delta">The amount of time to advance the clock by.</param>
     public void Advance(TimeSpan delta)
-        => SetUtcNow(_now + delta);
+    {
+        _ = Throw.IfLessThan(delta.Ticks, 0);
+
+        List<FakeTimeProviderTimer.Waiter>? waitersToWake;
+
+        lock (Waiters)
+        {
+            _now += delta;
+            waitersToWake = GetWaitersToWake();
+        }
+
+        WakeWaiters(waitersToWake);
```

Comment:
```suggestion
        SetUtcNow(_now + delta);
```

My instinct is to do this instead of duplicating the logic in both Advance and SetUtcNow. What is your thought process around not doing this?

Reply:
This would break atomicity and could lead to unexpected exceptions. If _now gets updated asynchronously after it is read here, the call to SetUtcNow could throw an exception because time is being set backward.

Reply:
This has been revamped.